File: scripts/bindings/ggplot_discrete_position_palette.py
"""FIX-GG04: materialized positional palettes through primary Python authoring."""
import json, math, sys
from pathlib import Path
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
out=Path(sys.argv[2]);out.mkdir(parents=True,exist_ok=True)
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes());options=c.export_options(640,360).dpi(96).basis('current')
cases=json.loads((ROOT/'fixtures/parity/ggplot2/discrete-position-palettes.json').read_text())['cases'];records=[]
def key(v):return 'Null' if v is None else {'Text':v}
def number(v):return {'number':'NaN' if v is None else v} if v is None or isinstance(v,str) else v
for index,case in enumerate(cases):
 for family in ('band','point'):
  for mode in (('primary',) if 'error' in case['result'] else ('primary','secondary')):
   owned=[]
   try:
    values=case['inputs'];expected=case['result'] if mode=='primary' else case['result']['secondary']
    data=c.Data.columns({'x':c.categorical([v or '' for v in values]).validity([v is not None for v in values]),'y':c.column([1.]*len(values),kind='float64')},keys=list(range(100,100+len(values))));owned.append(data)
    axis=c.x_axis().scale(c.scale_band() if family=='band' else c.scale_point()).range(100,540).discrete_policy({'limits':None if case['limits'] is None else list(map(key,case['limits'])),'palette':list(map(number,case['palette_values'])),'na_translate':case['na_translate']}).guide_geometry({'labels':'Preserve'})
    builder=c.plot(data).profile('Ggplot2_4_0_3').aes(c.aes().x('x').y('y')).layer(c.points()).x_axis(axis)
    if mode=='secondary':builder=builder.x_axis(c.x_axis().name('secondary').side('Top').secondary('x',1.,0.).guide_geometry({'labels':'Preserve'}))
    plot=builder.build();owned.append(plot);wire=plot.to_json();assert json.loads(wire)['version']==23
    restored=c.Plot.from_json(wire);owned.append(restored);assert restored.to_json()==wire
    request=output.request(restored,options);owned.append(request);frame=request.prepare();owned.append(frame);assert 'error' not in expected,(index,mode,expected)
    guides=frame.guides()['guides'];ticks=next(g for g in guides if g['spec']['side']==('Bottom' if mode=='primary' else 'Top'))['ticks']
    if mode=='secondary':
     assert len(ticks)==len(expected['breaks'])
     for tick,value,label,position in zip(ticks,expected['breaks'],expected['labels'],expected['positions']):
      assert math.isclose(tick['value']['Number'],value,rel_tol=0,abs_tol=1e-12)
      assert tick['label']==('NA' if label is None else label)
      assert math.isclose((tick['position']-100)/440,position,rel_tol=0,abs_tol=1e-12),(index,mode,tick,position)
    else:
     assert len(ticks)==sum(v is not None for v in expected['major_positions'])
     for value,label,position in zip(expected['breaks'],expected['labels'],expected['major_positions']):
      semantic='MissingCategory' if value is None else {'Category':value};found=[t for t in ticks if t['value']==semantic]
      if position is None:assert not found
      else:
       assert len(found)==1;tick=found[0];assert tick['label']==('NA' if label is None else label)
       assert math.isclose((tick['position']-100)/440,position,rel_tol=0,abs_tol=1e-12)
    scene=frame.scene();positions=[None]*len(values)
    for item,targets in zip(scene['items'],scene['targets']):
     if targets:positions[int(targets[0]['Source']['key'])-100]=(item['primitive']['Point']['center']['x']-100)/440
    for actual,wanted in zip(positions,case['result']['positions']):
     assert (actual is None)==(wanted is None),(index,mode,actual,wanted)
     if actual is not None:assert math.isclose(actual,wanted,rel_tol=0,abs_tol=1e-12),(index,mode,actual,wanted)
    records.append({'index':index,'family':family,'mode':mode,'ticks':ticks,'positions':positions})
    if family=='band' and (index,mode) in ((10,'secondary'),(11,'secondary'),(14,'secondary'),(17,'primary')):
     for fmt in ('svg','pdf','png'):(out/f'position-palette-{case["palette_name"]}.{fmt}').write_bytes(frame.export(fmt))
   except c.ChartError as error:
    diagnostic=json.loads(str(error));assert 'error' in expected,(index,mode,diagnostic)
    records.append({'index':index,'family':family,'mode':mode,'error':diagnostic['code']})
   finally:
    for value in reversed(owned):value.dispose()
assert len(records)==520
(out/'records.json').write_text(json.dumps(records,indent=2));options.dispose();output.dispose()
print('PASS Python: 520 positional palette configurations and twelve publication files.')

File: crates/chart-core/tests/ggplot_discrete_position_palette.rs
//! FIX-GG04 / GG2-03: fixed numeric positional palettes, independent R expectations.
use chart_core::{
    ChartResult, Rect, ResourceId, Revision, ScaleId, composition::ScaleValue, interpolate::Number,
    layout::*, prelude::*, scales::*, services::*,
};
use serde_json::Value;
fn cases() -> Vec<Value> {
    let fixture: Value = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/discrete-position-palettes.json"
    ))
    .unwrap();
    let cases = fixture["cases"].as_array().unwrap().clone();
    assert_eq!(cases.len(), 144);
    cases
}
fn key(v: &Value) -> ScaleKey {
    v.as_str()
        .map_or(ScaleKey::Null, |v| ScaleKey::Text(v.into()))
}
fn semantic(v: &Value) -> ScaleValue {
    v.as_str().map_or(ScaleValue::MissingCategory, |v| {
        ScaleValue::Category(v.into())
    })
}
fn number(v: &Value) -> f64 {
    match v.as_str() {
        Some("Infinity") => f64::INFINITY,
        Some("-Infinity") => f64::NEG_INFINITY,
        _ => v.as_f64().unwrap_or(f64::NAN),
    }
}
fn near(actual: f64, expected: f64, case: &Value) {
    assert!(
        actual == expected
            || (actual.is_nan() && expected.is_nan())
            || (actual - expected).abs() < 2e-14,
        "{actual} vs {expected}: {case}"
    );
}
fn projected(actual: Option<f64>, expected: &Value, case: &Value) {
    if expected.is_null() {
        assert!(actual.is_none(), "{actual:?} {case}")
    } else {
        near(actual.unwrap(), number(expected), case)
    }
}
fn policy(case: &Value) -> GgplotDiscretePosition {
    GgplotDiscretePosition {
        limits: case["limits"]
            .as_array()
            .map(|v| v.iter().map(key).collect()),
        palette: Some(
            case["palette_values"]
                .as_array()
                .unwrap()
                .iter()
                .map(|v| Number(number(v)))
                .collect(),
        ),
        na_translate: case["na_translate"].as_bool().unwrap(),
        ..Default::default()
    }
}
#[test]
fn materialized_position_palette_matches_reference_mapping_and_range() {
    for case in cases() {
        let values = case["inputs"]
            .as_array()
            .unwrap()
            .iter()
            .map(key)
            .collect::<Vec<_>>();
        let result = policy(&case).train(&values);
        let expected = &case["result"];
        if expected["error"].is_string() {
            assert!(result.is_err(), "{case}");
            continue;
        }
        let scale = result.unwrap_or_else(|e| panic!("{e:?} {case}"));
        for (a, e) in scale
            .viewport()
            .into_iter()
            .zip(expected["range"].as_array().unwrap())
        {
            near(a.0, number(e), &case)
        }
        for (i, key) in values.iter().enumerate() {
            projected(scale.map(key), &expected["mapped"][i], &case);
            projected(
                scale.project(key, Bounds::new(0., 1.).unwrap()).unwrap(),
                &expected["positions"][i],
                &case,
            );
        }
    }
}
struct Metrics;
impl TextMeasurer for Metrics {
    fn measure(&self, r: TextRequest<'_>) -> ChartResult<TextMetrics> {
        TextMetrics::new(r.text.len() as f64 * 5., 8., 2.)
    }
}
#[test]
fn primary_palette_and_secondary_guides_share_mapped_category_coordinates() {
    let request = LayoutRequest::new(
        Rect::new(0., 0., 640., 360.).unwrap(),
        Units::LogicalPixels,
        ResourceDescriptor {
            id: ResourceId::new(1),
            revision: Revision::INITIAL,
            kind: ResourceKind::Font,
            byte_len: 1,
        },
    );
    for case in cases() {
        for point in [false, true] {
            let values = case["inputs"].as_array().unwrap();
            let data = Data::columns()
                .column(
                    "x",
                    categorical(values.iter().map(|v| v.as_str().unwrap_or("")))
                        .validity(values.iter().map(|v| !v.is_null()).collect()),
                )
                .column("y", vec![1.; values.len()])
                .build()
                .unwrap();
            let builder = plot(data)
                .profile(Profile::Ggplot2_4_0_3)
                .aes(aes().x("x").y("y"))
                .layer(points())
                .x_axis(
                    x_axis()
                        .scale(if point { scale_point() } else { scale_band() })
                        .range(0., 100.)
                        .discrete_policy(Some(policy(&case)))
                        .guide_geometry(Some(GuideGeometry {
                            labels: Some(GuideLabelPolicy::Preserve),
                            ..Default::default()
                        })),
                );
            let figure = builder.clone().build().unwrap();
            assert_eq!(figure.definition().wire_version(), 23);
            let wire = figure.to_json().unwrap();
            let figure = Plot::from_json(&wire).unwrap();
            let mut stale: Value = serde_json::from_str(&wire).unwrap();
            stale["version"] = 22.into();
            assert!(Plot::from_json(&stale.to_string()).is_err());
            let result = layout(
                figure.chart().unwrap().prepare().unwrap(),
                &request,
                &Metrics,
            );
            let expected = &case["result"];
            if expected["error"].is_string() {
                assert!(result.is_err(), "{case}");
                continue;
            }
            let frame = result.unwrap_or_else(|e| panic!("{e:?} {case}"));
            let axis = &frame.axes()[&ScaleId::new(0)];
            for (i, value) in values.iter().enumerate() {
                projected(
                    axis.map_value(&semantic(value)).unwrap().map(|p| p / 100.),
                    &expected["positions"][i],
                    &case,
                )
            }
            for (i, key) in expected["breaks"].as_array().unwrap().iter().enumerate() {
                let tick = axis.ticks.iter().find(|tick| tick.value == semantic(key));
                let position = &expected["major_positions"][i];
                if position.is_null() {
                    assert!(tick.is_none(), "{case}")
                } else {
                    let tick = tick.unwrap();
                    near(tick.position / 100., number(position), &case);
                    assert_eq!(
                        tick.label,
                        expected["labels"][i].as_str().unwrap_or("NA"),
                        "{case}"
                    )
                }
            }
            let secondary = builder
                .x_axis(
                    x_axis()
                        .name("secondary")
                        .side(AxisSide::Top)
                        .secondary("x", 1., 0.)
                        .guide_geometry(Some(GuideGeometry {
                            labels: Some(GuideLabelPolicy::Preserve),
                            ..Default::default()
                        })),
                )
                .build()
                .unwrap();
            let id = secondary.axis("secondary").unwrap().id();
            let result = layout(
                secondary.chart().unwrap().prepare().unwrap(),
                &request,
                &Metrics,
            );
            let expected = &expected["secondary"];
            if expected["error"].is_string() {
                assert!(result.is_err(), "{case}");
                continue;
            }
            let frame = result.unwrap_or_else(|e| panic!("{e:?} {case}"));
            let axis = &frame.axes()[&id];
            assert_eq!(
                axis.ticks.len(),
                expected["breaks"].as_array().unwrap().len(),
                "{case}"
            );
            for (i, tick) in axis.ticks.iter().enumerate() {
                let ScaleValue::Number(value) = tick.value else {
                    panic!()
                };
                near(value, number(&expected["breaks"][i]), &case);
                near(
                    tick.position / 100.,
                    number(&expected["positions"][i]),
                    &case,
                );
                assert_eq!(
                    tick.label,
                    expected["labels"][i].as_str().unwrap_or("NA"),
                    "{case}"
                )
            }
        }
    }
}

#[test]
fn fixed_position_palettes_retrain_after_replacement_without_mutating_captures() {
    let request = LayoutRequest::new(
        Rect::new(0., 0., 640., 360.).unwrap(),
        Units::LogicalPixels,
        ResourceDescriptor {
            id: ResourceId::new(1),
            revision: Revision::INITIAL,
            kind: ResourceKind::Font,
            byte_len: 1,
        },
    );
    let data = |values: &[Option<&str>]| {
        Data::columns()
            .column(
                "x",
                categorical(values.iter().map(|v| v.unwrap_or("")))
                    .validity(values.iter().map(Option::is_some).collect()),
            )
            .column("y", vec![1.; values.len()])
            .build()
            .unwrap()
    };
    let layer = points();
    let secondary_axis = x_axis()
        .name("secondary")
        .side(AxisSide::Top)
        .secondary("x", 1., 0.);
    let build = |data: Data| {
        plot(data)
            .profile(Profile::Ggplot2_4_0_3)
            .aes(aes().x("x").y("y"))
            .layer(layer.clone())
            .x_axis(x_axis().discrete_policy(Some(GgplotDiscretePosition {
                palette: Some(vec![8_f64.into(), 2_f64.into(), 5_f64.into(), 3_f64.into()]),
                ..Default::default()
            })))
            .x_axis(secondary_axis.clone())
            .build()
            .unwrap()
    };
    let original_data = data(&[Some("a"), None, Some("b")]);
    let original = build(original_data.clone());
    let wire = original.to_json().unwrap();
    let mut updated = original.chart().unwrap();
    let held = updated.prepare().unwrap();
    let initial = layout(held.clone(), &request, &Metrics)
        .unwrap()
        .guide_snapshots();
    for values in [
        &[Some("c"), None, Some("b"), Some("a")][..],
        &[Some("b"), None][..],
        &[][..],
        &[Some("a"), None, Some("b")][..],
    ] {
        let transaction = updated
            .transaction()
            .unwrap()
            .replace(&original_data, data(values))
            .build()
            .unwrap();
        updated.apply_transaction(transaction).unwrap();
        let revised = layout(updated.prepare().unwrap(), &request, &Metrics).unwrap();
        let fresh = layout(
            build(data(values)).chart().unwrap().prepare().unwrap(),
            &request,
            &Metrics,
        )
        .unwrap();
        assert_eq!(revised.guide_snapshots(), fresh.guide_snapshots());
        assert_eq!(revised.scene().items(), fresh.scene().items());
        assert_eq!(
            layout(held.clone(), &request, &Metrics)
                .unwrap()
                .guide_snapshots(),
            initial
        );
        assert_eq!(original.to_json().unwrap(), wire);
    }
}
